### PR TITLE
Implement a validation pass to fix one-frame layout issues

### PR DESCRIPTION
**Pull request by smoogipoo**
Fixes ppy/osu#3136

At the moment this implementation serves two purposes: it provides the necessary interfaces to validate layouts, and helps resolve some simple cases due to layout validation order.

#### Drawable.cs
```diff
+ public virtual bool RequiresLayoutValidation { get; }

+ protected virtual void ValidateLayout()
+ public virtual bool ValidateSubTree()
```

This adds a new "validation" pass to the draw graph, which runs post-`UpdateSubTree`/post-`UpdateSubTreeMasking`. This occurs through the recursive `ValidateSubTree` method.

`RequiresLayoutValidation`
  * Determines if layout needs to be determined.
  * Should basically always refer to the `IsValid` state of some `Cached` value(s).

`ValidateLayout()`
  * Performs validation of the layout, validating any values which `RequiresLayoutValidation` refers to.
  * Replaces code which was previously present inside `UpdateAfterChildren`.

The pass will run continuously on sub-trees until `RequiresLayoutValidation` becomes `false` for all children, up to a maximum of 10 iterations. I haven't seen more than 1 iteration ever run as autosize isn't validated, but if autosize validations were performed, this will iterate anywhere from 2-4 times in an average case. It is currently problematic to validate autosize due to the implementation of `BoundingBox`.

A WIP version with autosize validations (and a fix to `BoundingBox`) can be found at https://github.com/smoogipoo/osu-framework/tree/autosize-validations . Which also fixes cases like the one-frame issue on score overlay.

### CODE REVIEW

**Comment by smoogipoo** [comment]
Testing could still be improved, but I think it's sufficient for now? Hoping we can add more _after_ we see additional failure cases.

File: osu.Framework/Graphics/Containers/CompositeDrawable.cs [hunk 1]
```diff
@@ -554,6 +556,45 @@ public override bool UpdateSubTreeMasking(Drawable source, RectangleF maskingBou
...
+            // This should never happen, but exists for safety purposes.
+            const int max_iterations = 10;
+
+            if (!IsPresent)
+                return true;
+
+            int validations = 0;
+
+            bool moreRequired = true;
+            while (moreRequired && validations < max_iterations)
```
Comment (Tom94) [inline osu.Framework/Graphics/Containers/CompositeDrawable.cs]: `for (; validations < max_iterations && moreRequired; ++validations)`?

Comment (Tom94) [reply osu.Framework/Graphics/Containers/CompositeDrawable.cs]: Also, perhaps consider removing `moreRequired` and replacing it with a local bool within the loop, combined with `break`?

**Comment by peppy** [comment]
Also needs conflict resolution

**Comment by smoogipoo** [comment]
Closing as this has been superseded by @ekrctb 's PRs.

**Pull request closed without merging**